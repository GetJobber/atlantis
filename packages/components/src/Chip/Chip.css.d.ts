declare const styles: {
  readonly "chip": string;
  readonly "prefix": string;
  readonly "suffix": string;
  readonly "empty": string;
  readonly "chipBar": string;
  readonly "selected": string;
  readonly "base": string;
  readonly "invalid": string;
  readonly "subtle": string;
  readonly "disabled": string;
};
export = styles;

