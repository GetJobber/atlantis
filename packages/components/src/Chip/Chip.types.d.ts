import React, { PropsWithChildren } from "react";

export interface ChipProps extends PropsWithChildren {
  /** Accessible Label */
  ariaLabel?: string;
  /** Disables Chip */
  disabled?: boolean;
  /** Adds more prominent text to act as a heading */
  heading?: string;
  /** Changes Chip styling to inform the user of an issue. */
  invalid?: boolean;
  /** What do you want Chip to say? */
  label: string;
  /** What role does Chip play? */
  role?: string;
  /** Defaults to zero. Accessible feature. */
  tabIndex?: number;
  /** Defaults to base. Changes style of button */
  variation?: "subtle" | "base";
  /** Chip Click Callback. Sends an event and sometimes a value (SelectableChip).  */
  onClick?: (
    ev: React.MouseEvent<HTMLButtonElement>,
    value?: string | number,
  ) => void;
  /**  Callback. Called when you keydown on Chip. Ships the event, so you can get the key pushed.  */
  onKeyDown?: (ev: React.KeyboardEvent<HTMLButtonElement>) => void;
}

export interface ChipSelectableProps extends ChipProps {
  /** Is Chip selected? */
  selected?: boolean;
  /** Send this back onClick. Good for determining which Chip was clicked. */
  value?: string | number;
}
