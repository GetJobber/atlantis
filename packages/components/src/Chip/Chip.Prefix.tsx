import React, { PropsWithChildren } from "react";
import styles from "./Chip.css";

export function ChipPrefix({ children }: PropsWithChildren) {
  return <span className={styles.prefix}>{children}</span>;
}
