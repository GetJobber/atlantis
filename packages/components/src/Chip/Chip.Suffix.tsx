import React, { PropsWithChildren } from "react";
import classNames from "classnames";
import { useChildComponent } from "@jobber/hooks/useChildComponent";
import styles from "./Chip.css";
import { Icon } from "../Icon";

interface ChipSuffixProps extends PropsWithChildren {
  className?: string;
}

export function ChipSuffix({ children, className }: ChipSuffixProps) {
  let singleChild = useChildComponent(children, d => d.type === Icon);
  if (!["cross", "add", "checkmark"].includes(singleChild?.props?.name)) {
    singleChild = undefined;
  }
  return (
    <span
      className={classNames(
        styles.suffix,
        className,
        !singleChild && styles.empty,
      )}
    >
      {singleChild}
    </span>
  );
}
