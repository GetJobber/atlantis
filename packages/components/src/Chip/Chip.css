.chip {
  --chip-height: 40px;
  --chip-radius: 20px;
  --chip-disabled--bg: var(--color-grey--lighter);
  --chip-bar-color: var(--color-greyBlue--light);
  --chip-base-bg: var(--color-greyBlue--lightest);
  --chip-base--hover: var(--color-greyBlue--lighter);
  display: flex;
  position: relative;
  height: var(--chip-height);
  border: var(--border-base) solid transparent;
  border-radius: var(--chip-radius);
  color: var(--color-heading);
  text-align: left;
  cursor: pointer;
  transition: all var(--ease-both) var(--timing-quick);
  align-items: center;
}
div.chip {
  padding: 0 var(--base-unit);
}
button.chip {
  padding: var(--space-small) var(--base-unit);
}
.chip.actingAsDiv {
  cursor: auto;
}

.chip:focus {
  box-shadow: var(--shadow-focus);
}

.chip .prefix {
  margin-left: calc(-1 * var(--space-small));
  margin-right: var(--space-small);
}

.chip .suffix {
  margin-left: var(--space-small);
  margin-right: calc(-1 * var(--space-small));
}

.chip .prefix,
.chip .suffix {
  display: flex;
  width: var(--space-large);
  height: var(--space-large);
  flex-shrink: 0;
  border-radius: var(--radius-circle);
  background-color: var(--color-surface);
  align-items: center;
  justify-content: center;
}

.chip .suffix:hover,
.chip .prefix:hover {
  background-color: var(--color-surface--hover);
}

.chip .prefix.empty,
.chip .suffix.empty {
  display: none;
}

.chip .chipBar {
  width: var(--space-minuscule);
  height: var(--base-unit);
  margin: 0 var(--space-small);
  background-color: var(--chip-bar-color);
}

.chip .selected {
  border-radius: var(--radius-circle);
  background-color: var(--color-surface);
}

.base {
  border: var(--border-base) solid transparent;
  color: var(--color-heading);
  background-color: var(--chip-base-bg);
}
.base:hover,
.base:focus {
  background-color: var(--chip-base--hover);
}

.invalid,
.invalid:hover,
.invalid:focus,
.subtle.invalid,
.subtle.invalid:hover,
.subtle.invalid:focus {
  border: var(--border-base) solid var(--color-critical);
  background-color: var(--color-critical--surface);
}

.subtle {
  border: var(--border-base) solid var(--color-disabled--secondary);
  background-color: var(--color-surface);
}
.subtle:hover {
  border: var(--border-base) solid var(--chip-base-bg);
  background-color: var(--chip-base-bg);
}

.disabled,
.disabled:hover,
.disabled:focus {
  border: var(--border-base) solid var(--chip-disabled--bg);
  color: var(--color-disabled);
  background-color: var(--chip-disabled--bg);
  cursor: auto;
}

.disabled .chipBar {
  background-color: var(--color-disabled);
}

.chip.disabled .suffix {
  background-color: var(--color-disabled);
}

.clickableSuffix {
  cursor: pointer;
}
