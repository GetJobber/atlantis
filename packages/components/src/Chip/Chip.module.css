:root {
  --public-chip-base-bg-color: var(--color-interactive--background);
  --public-chip-base-hover-bg-color: var(
    --color-interactive--background--hover
  );
  --public-chip-base-content-color: var(--color-heading);
}

.chip {
  --chip-base-bg-color: var(--public-chip-base-bg-color);
  --chip-base-hover-bg-color: var(--public-chip-base-hover-bg-color);
  --chip-base-content-color: var(--public-chip-base-content-color);
  --chip-border-color: transparent;
  --chip-height: 40px;
  --chip-radius: 20px;
  display: inline-flex;
  position: relative;
  height: var(--chip-height);
  min-width: 0;
  box-sizing: border-box;
  padding: 0 var(--base-unit);
  border: var(--border-base) solid var(--chip-border-color);
  border-radius: var(--chip-radius);
  color: var(--color-heading);
  text-align: left;
  background-color: var(--chip-base-bg-color);
  cursor: pointer;
  transition: all var(--timing-base);
  align-items: center;
}

.chip:hover,
.chip:focus,
.chip:focus-visible,
.clickableSuffix:focus-visible {
  border: var(--border-base) solid var(--chip-border-color);
  outline: transparent;
  background-color: var(--chip-base-hover-bg-color);
}

.chip .clickableSuffix {
  padding: 0;
  border: none;
  cursor: pointer;
}

.chip .suffix,
.chip .clickableSuffix {
  margin-left: var(--space-small);
  margin-right: calc(-1 * var(--space-small));
  background-color: var(--color-surface);
}

.chip .prefix,
.chip .suffix,
.chip .clickableSuffix {
  display: flex;
  width: var(--space-large);
  height: var(--space-large);
  flex-shrink: 0;
  border-radius: var(--radius-circle);
  align-items: center;
  justify-content: center;
}

.clickableSuffix:hover {
  background-color: var(--color-surface--hover);
}

.chip:focus-visible,
.chip .clickableSuffix:focus-visible {
  box-shadow: var(--shadow-focus);
}

.chip .prefix {
  margin-left: calc(-1 * var(--space-small));
  margin-right: var(--space-small);
}

.chip .prefix.empty,
.chip .suffix.empty {
  display: none;
}

.chip .chipBar {
  width: var(--space-minuscule);
  height: var(--base-unit);
  margin: 0 var(--space-small);
  background-color: var(--color-text--secondary);
}

.chip .selected {
  border-radius: var(--radius-circle);
  background-color: var(--color-surface);
}

.base,
.base p {
  color: var(--chip-base-content-color);
}

.base .chipBar {
  background-color: var(--chip-base-content-color);
}

.invalid,
.subtle.invalid {
  --chip-border-color: var(--color-critical);
  --chip-base-bg-color: var(--color-critical--surface);
  --chip-base-hover-bg-color: var(--color-interactive--background--hover);
}

/* If these change, please update InternalChip.css focus-visible rules */
.invalid:hover,
.invalid:focus-visible,
.subtle.invalid:hover,
.subtle.invalid:focus-visible {
  --chip-border-color: var(--color-critical);
}

.subtle {
  --chip-border-color: var(--color-border--interactive);
  --chip-base-bg-color: var(--color-surface);
  --chip-base-hover-bg-color: var(--color-interactive--background);
}

/* If these change, please update InternalChip.css focus-visible rules */
.subtle:focus-visible,
.subtle:hover {
  --chip-border-color: var(--chip-base-bg-color);
}

.invalid.disabled,
.invalid.disabled:hover,
.invalid.disabled:focus-visible {
  --chip-base-bg-color: var(--color-disabled--secondary);
  --chip-base-hover-bg-color: var(--color-disabled--secondary);
  cursor: not-allowed;
}

/* If these change, please update InternalChip.css focus-visible rules */
.disabled,
.disabled:hover,
.disabled:focus-visible {
  --chip-border-color: var(--color-disabled--secondary);
  color: var(--color-disabled);
  --chip-base-bg-color: var(--color-disabled--secondary);
  --chip-base-hover-bg-color: var(--color-disabled--secondary);
  cursor: not-allowed;
}

.disabled .chipBar {
  background-color: var(--color-disabled);
}

.chip.disabled .suffix {
  background-color: var(--color-disabled--secondary);
}

.chip.disabled .suffix path {
  fill: var(--color-disabled);
}

.chipContent {
  display: flex;
  position: relative;
  overflow: hidden;
  flex-wrap: nowrap;
}

.chipContent > * {
  flex: 0 0 auto;
}

.chipContent > p {
  overflow: hidden;
  white-space: nowrap;
}

.truncateGradient {
  position: absolute;
  top: 0;
  right: 0;
  width: var(--space-base);
  height: 100%;
}

.truncateGradient::before,
.truncateGradient::after {
  content: "";
  position: absolute;
  inset: 0;
  transition: opacity var(--timing-base);
}

.truncateGradient::before {
  background: linear-gradient(
    to left,
    var(--chip-base-bg-color) 5%,
    rgba(255, 255, 255, 0)
  );
  opacity: 1;
}

.truncateGradient::after {
  background: linear-gradient(
    to left,
    var(--chip-base-hover-bg-color) 5%,
    rgba(255, 255, 255, 0)
  );
  opacity: 0;
}

.chip:hover .truncateGradient::before,
.chip:focus-visible .truncateGradient::before {
  opacity: 0;
}

.chip:hover .truncateGradient::after,
.chip:focus-visible .truncateGradient::after {
  opacity: 1;
}

.chipLabelRef {
  display: inline-block;
  width: 1px;
}
