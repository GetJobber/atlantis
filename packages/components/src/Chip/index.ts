export * from "./Chip";
export * from "./components/ChipSelectable/ChipSelectable";
export * from "./components/ChipDismissible/ChipDismissible";
export * from "./Chip.types";
