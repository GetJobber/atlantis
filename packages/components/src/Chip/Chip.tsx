import React from "react";
import { useChildComponent } from "@jobber/hooks/useChildComponent";
import classNames from "classnames";
import styles from "./Chip.css";
import { ChipPrefix } from "./components/ChipPrefix/Chip.Prefix";
import { ChipSuffix } from "./components/ChipSuffix/Chip.Suffix";
import { ChipProps } from "./Chip.types";
import { Typography } from "../Typography";

export const Chip = ({
  ariaLabel,
  disabled,
  heading,
  invalid,
  label,
  onClick,
  onKeyDown,
  children,
  role = "button",
  tabIndex = 0,
  variation = "base",
}: ChipProps): JSX.Element => {
  const classes = classNames(styles.chip, {
    [styles.invalid]: invalid,
    [styles.base]: variation === "base",
    [styles.subtle]: variation === "subtle",
    [styles.disabled]: disabled,
  });

  const prefix = useChildComponent(children, d => d.type === Chip.Prefix);
  const suffix = useChildComponent(children, d => d.type === Chip.Suffix);

  return (
    <button
      className={classes}
      onClick={onClick}
      tabIndex={tabIndex}
      onKeyDown={onKeyDown}
      aria-label={ariaLabel || label}
      disabled={disabled}
      role={role}
    >
      {prefix}
      <span>
        <Typography size="base" fontWeight="medium">
          {heading}
        </Typography>
      </span>
      {heading && <span className={styles.chipBar}></span>}
      <span>
        <Typography size="base">{label}</Typography>
      </span>
      {suffix}
    </button>
  );
};

Chip.Prefix = ChipPrefix;
Chip.Suffix = ChipSuffix;
