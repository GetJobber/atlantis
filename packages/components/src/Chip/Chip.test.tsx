import React from "react";
import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Chip } from "./Chip";
import styles from "./Chip.module.css";
import { Avatar } from "../Avatar";

let mockIsInView = jest.fn(() => true);
jest.mock("@jobber/hooks/useInView", () => ({
  useInView: () => [{ current: null }, mockIsInView()],
}));

describe("Chip", () => {
  it("renders without error", () => {
    const onClick = jest.fn();
    const { getByText } = render(<Chip label="Label" onClick={onClick} />);

    expect(getByText("Label")).toBeInTheDocument();
  });

  it("calls onClick when clicked", () => {
    const onClick = jest.fn();
    const { getByText } = render(<Chip onClick={onClick} label="Test Chip" />);

    fireEvent.click(getByText("Test Chip"));

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("calls onKeyDown when typed in", () => {
    const onKeyDown = jest.fn();
    const { getByText } = render(
      <Chip onKeyDown={onKeyDown} label="Test Chip" />,
    );
    fireEvent.keyDown(getByText("Test Chip"), { key: "Enter" });
    expect(onKeyDown).toHaveBeenCalledTimes(1);
  });

  it("does not fire onClick when disabled", () => {
    const label = "Test Disabled";
    const clickHandler = jest.fn();
    const { getByText } = render(
      <Chip label={label} onClick={clickHandler} disabled />,
    );

    fireEvent.click(getByText(label));
    expect(clickHandler).toHaveBeenCalledTimes(0);
  });

  it("should have a role of button when role not provided", () => {
    const { getByRole } = render(<Chip label="Test Chip" />);
    expect(getByRole("button")).toBeInstanceOf(HTMLDivElement);
  });

  it("should have a tabIndex of zero if one is not provided", () => {
    const { getByRole } = render(<Chip label="Test Chip" />);

    expect(getByRole("button")).toHaveAttribute("tabIndex", "0");
    expect(getByRole("button")).toBeInstanceOf(HTMLDivElement);
  });

  it("should set aria-label if one is provided", () => {
    const { getByRole } = render(
      <Chip label="Test Chip" ariaLabel="Test Label" />,
    );

    expect(getByRole("button")).toHaveAttribute("aria-label", "Test Label");
  });
});

describe("Chip Children", () => {
  it("should only render one prefix if two are provided", () => {
    const { container } = render(
      <Chip label="Test Chip">
        <Chip.Prefix>
          <Avatar initials="" />
        </Chip.Prefix>
        <Chip.Prefix>
          <Avatar initials="" />
        </Chip.Prefix>
      </Chip>,
    );
    const elements = container.querySelectorAll("." + styles.prefix);

    expect(elements).toHaveLength(1);
  });

  it("should only render one suffix if two are provided", () => {
    const { container } = render(
      <Chip label="Test Chip">
        <Chip.Suffix>
          <Avatar initials="" />
        </Chip.Suffix>
        <Chip.Suffix>
          <Avatar initials="" />
        </Chip.Suffix>
      </Chip>,
    );
    const elements = container.querySelectorAll("." + styles.suffix);

    expect(elements).toHaveLength(1);
  });
});

describe("Chip truncation", () => {
  describe("with only a label too long to display", () => {
    beforeEach(() => {
      mockIsInView.mockReturnValueOnce(false);
    });

    it("should show a tooltip with the label's content", async () => {
      const label =
        "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
      const { getByRole } = render(<Chip label={label} />);
      const button = getByRole("button");

      await userEvent.hover(button);

      expect(document.querySelector("div[role='tooltip']")).toHaveTextContent(
        label,
      );
    });

    it("should show a gradient to indicate truncation", () => {
      const label =
        "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
      const { getByTestId } = render(<Chip label={label} />);
      const gradient = getByTestId("ATL-Chip-Truncation-Gradient");

      expect(gradient).toBeInTheDocument();
    });
  });

  describe("with only a label short enough to display", () => {
    beforeEach(() => {
      mockIsInView = jest.fn(() => true);
    });

    it("should not show a tooltip", async () => {
      const label = "short";
      const { getByRole } = render(<Chip label={label} />);
      const button = getByRole("button");

      await userEvent.hover(button);

      expect(
        document.querySelector("div[role='tooltip']"),
      ).not.toBeInTheDocument();
    });

    it("should not show a gradient", () => {
      const label = "short";
      const { queryByTestId } = render(<Chip label={label} />);
      const gradient = queryByTestId("ATL-Chip-Truncation-Gradient");

      expect(gradient).not.toBeInTheDocument();
    });
  });

  describe("with a label and heading", () => {
    describe("when only label is too long to display", () => {
      beforeEach(() => {
        mockIsInView.mockReturnValueOnce(false);
      });

      it("should show a tooltip with only the label content", async () => {
        const heading = "heading";
        const label =
          "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
        const { getByRole } = render(<Chip label={label} heading={heading} />);
        const button = getByRole("button");

        await userEvent.hover(button);

        expect(document.querySelector("div[role='tooltip']")).toHaveTextContent(
          label,
        );
      });

      it("should show a gradient to indicate truncation", () => {
        const label =
          "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
        const { getByTestId } = render(<Chip label={label} />);
        const gradient = getByTestId("ATL-Chip-Truncation-Gradient");

        expect(gradient).toBeInTheDocument();
      });
    });

    describe("when heading is also too long to display", () => {
      beforeEach(() => {
        mockIsInView.mockReturnValue(false);
      });

      it("should show a tooltip with the label and heading content", async () => {
        const heading = "heading";
        const label =
          "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
        const { getByRole } = render(<Chip label={label} heading={heading} />);
        const button = getByRole("button");

        await userEvent.hover(button);

        expect(document.querySelector("div[role='tooltip']")).toHaveTextContent(
          `${heading} | ${label}`,
        );
      });

      it("should show a gradient to indicate truncation", () => {
        const label =
          "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
        const { getByTestId } = render(<Chip label={label} />);
        const gradient = getByTestId("ATL-Chip-Truncation-Gradient");

        expect(gradient).toBeInTheDocument();
      });
    });

    describe("when both can fully display", () => {
      beforeEach(() => {
        mockIsInView.mockReturnValue(true);
      });

      it("should not show a tooltip on hover", async () => {
        const heading = "heading";
        const label =
          "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
        const { getByRole } = render(<Chip label={label} heading={heading} />);
        const button = getByRole("button");

        await userEvent.hover(button);

        expect(document.querySelector("div[role='tooltip']")).toBeNull();
      });

      it("shold now show a gradient to indicate truncation", () => {
        const label =
          "Has Anyone Really Been Far Even as Decided to Use Even Go Want to do Look More Like?";
        const { queryByTestId } = render(<Chip label={label} />);
        const gradient = queryByTestId("ATL-Chip-Truncation-Gradient");

        expect(gradient).not.toBeInTheDocument();
      });
    });
  });
});
