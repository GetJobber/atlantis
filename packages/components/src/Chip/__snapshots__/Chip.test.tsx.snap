// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip renders a disabled Chip 1`] = `
<div>
  <button
    aria-label="Test Chip"
    class="chip base disabled"
    disabled=""
    role="button"
    tabindex="0"
  >
    <span>
      <p
        class="base medium base"
      />
    </span>
    <span>
      <p
        class="base regular base"
      >
        Test Chip
      </p>
    </span>
  </button>
</div>
`;

exports[`Chip renders a disabled and invalid Chip 1`] = `
<div>
  <button
    aria-label="Test Chip"
    class="chip invalid base disabled"
    disabled=""
    role="button"
    tabindex="0"
  >
    <span>
      <p
        class="base medium base"
      />
    </span>
    <span>
      <p
        class="base regular base"
      >
        Test Chip
      </p>
    </span>
  </button>
</div>
`;

exports[`Chip renders an invalid Chip 1`] = `
<div>
  <button
    aria-label="Test Chip"
    class="chip invalid base"
    role="button"
    tabindex="0"
  >
    <span>
      <p
        class="base medium base"
      />
    </span>
    <span>
      <p
        class="base regular base"
      >
        Test Chip
      </p>
    </span>
  </button>
</div>
`;
